Feature/select process

csadorf opened the pull request:
Enable the switch between previously submitted processes and the submission of new processes.

yakutovicha commented on process.py:
```diff
@@ -150,5 +156,146 @@ def __init__(self, **kwargs):
+
+    # When this trait is set to a positive value, the work chains are automatically
+    # refreshed every `auto_refresh_interval` seconds.
+    auto_refresh_interval = traitlets.Int(allow_none=True)  # seconds
```
will it work if you set it to `None`?

csadorf replied:
I don't think so, good catch!
```suggestion
    auto_refresh_interval = traitlets.Int()  # seconds
```

yakutovicha reviewed (commented):
Thanks, @csadorf, it is a very good feature. I tried different scenarios of running the app. Based on my interaction experience I have a few questions/suggestions:
- Is it possible to call the reset, when the structure is changed? I believe you can observe the `structure` trait for that.
- Related to the previous point. If, after changing the structure, I click on the refresh button - the structure remains unchanged.
- If I open the "Select structure" accordion, the `reset` button gets deactivate. Is it intended?
- If, after selecting a finished calculation, I go back to the `New calculation..` and click on `Refresh` button - the structure remains unchanged. If I change the structure, the `Confirm` button remains inactive. Only after pressing `RESET` the `confirm` button gets activated.
- Selecting the running calculation seems not to be very responsive.

csadorf commented on bands.py:
```diff
@@ -43,20 +43,49 @@ def _observe_input_structure(self, change):
+            # loaded are consistent.
+            assert process_node.process_label == 'PwBandsWorkChain'
+            assert builder.scf.pseudo_family.value == builder.bands.pseudo_family.value
+            assert dict(load_default_parameters()) \
+                    == dict(builder.scf.pw['parameters']) \
+                    == dict(builder.bands.pw['parameters'])
+            assert builder.scf.kpoints_distance == Float(0.8)
```
This check is used to ensure that the parameters of the process we load is the same that we would use to submit. Otherwise a very basic user assumption would be broken, namely that resubmitting a previous process is actually submitting with the same parameters. It would probably be even better to only show processes in the list with those parameters.

yakutovicha replied:
Maybe I don't entirely understand the motivation behind but to me, a slight modification of the input parameters would be normal. For instance, for the metals, you want to include smearing, while for non-metals you don't. I think the "clever" work chains actually do those changes. So I am still not fully convinced this check is needed.

csadorf replied:
If you load a process and then just resubmit it to a different host. Wouldn't you expect it to run the exact same calculation? Or maybe even on the same host? Would you not be very surprised if the calculation executed ends up actually being a different one? I would be. 😕

csadorf commented on codes.py:
```diff
@@ -241,38 +244,47 @@ def _refresh_outputs_view(self, change=None):
-        assert self.process is not None
-        process_node = load_node(self.process.id)
+    def _monitor_process(self, process_id, timeout=.1):
+        self._monitor_thread_stop.wait(timeout=10 * timeout)  # brief delay to increase app stability
+
+        process = None if process_id is None else load_node(process_id)
+
+        iteration = itertools.count()
+        while not (process is None or process.is_sealed):
```
All I know is that without a brief delay at that point the app does not load anymore. At least on my system in my browser. I do not know the cause. The correctness of the app behavior is not affected, it just slows down the initial load by a second.

I am not quire sure what you mean by "break it".

yakutovicha replied:
> I am not quire sure what you mean by "break it".

By "break it" I meant that the delays that can be triggered by some external conditions (out of RAM memory, slow or overloaded hard drive) can result in some unexpected behaviour.

csadorf replied:
The behavior is already unexpected since I don't know what is causing the problem.

csadorf commented:
> Thanks, @csadorf, it is a very good feature. 

Thx!

I tried different scenarios of running the app. Based on my interaction experience I have a few questions/suggestions:
> 
>     * Is it possible to call the reset, when the structure is changed? I believe you can observe the `structure` trait for that.

Ideally it should not even be possible to select a different structure without resetting, but I'll look into that.

>     * Related to the previous point. If, after changing the structure, I click on the refresh button - the structure remains unchanged.

The "refresh" button refreshes the list of work chains, not the selected work chain. That is probably ambiguous. Since the list of work chains is auto-refreshed every 10 seconds or so anyways, should we maybe just remove the "refresh" button?
> 
>     * If I open the "Select structure" accordion, the `reset` button gets deactivate. Is it intended?
> 

Yes, you can only reset the latest locked step, in that case it is step 2.

>     * If, after selecting a finished calculation, I go back to the `New calculation..` and click on `Refresh` button - the structure remains unchanged. If I change the structure, the `Confirm` button remains inactive. Only after pressing `RESET` the `confirm` button gets activated.
> 

This seems like a bug in the structure selection step. Like I said before, it should not be possible to change the structure once it is confirmed. I'll look into it.

>     * Selecting the running calculation seems not to be very responsive.

What exactly is not responsive? I've tried to streamline the implementation as much as possible and am not sure how to make it more efficient, but maybe you can describe to me what exactly you would expect to happen faster and I can double check for possible issues.

csadorf commented:
@yakutovicha I believe I've addressed all of your comments. I've changed the app behavior in the way that you suggested. So clicking reset on step 1 will actually reset all steps and furthermore, selecting a new structure automatically resets the app. Let me know what you think!

yakutovicha reviewed (commented):
Thanks @csadorf.

I tested the newest changes and I have three comments:

1. Looks like there is a loop that presses the `Refresh` button with a certain interval. I don't know whether this is only my case but after a certain number of auto-refreshes, the latest output gets frozen.
2. Another thing is related to the code selection widget. I realized that if the code, which was used for the finished calculation, is not available - the app fails. This is entirely the problem of the `CodeDropdown` widget and has been fixed in the previous release. This makes it important to update the image before running the tutorial.
3. I would probably remove the structure after selecting the "New calculation" in the process selection widget, but this is up to you in the end.

csadorf commented:
>     1. Looks like there is a loop that presses the `Refresh` button with a certain interval. I don't know whether this is only my case but after a certain number of auto-refreshes, the latest output gets frozen.
> 

No, I have not experienced that issue. Does the widget simply not update anymore or does it remain disabled?

>     2. Another thing is related to the code selection widget. I realized that if the code, which was used for the finished calculation, is not available - the app fails. This is entirely the problem of the `CodeDropdown` widget and has been fixed in the previous release. This makes it important to update the image before running the tutorial.

👍 

>     3. I would probably remove the structure after selecting the "New calculation" in the process selection widget, but this is up to you in the end.

I'll consider it.

csadorf commented:
@yakutovicha I've played around with resetting the structure as well and found that it leads to inconsistencies to what the "New calculation..." state implies. So for the time being, I'd prefer to leave it _as is_.

I have not been able to identify any issues with the process selection auto-refresh.

csadorf commented:
@yakutovicha Issue should be fixed now.

yakutovicha reviewed (approved):
thanks, @csadorf .

There is one little remaining issue:  if one runs a calculation, then switches to a completed one, and then switches back to the running calculation - the on-the-fly output gets frozen.

But I think it is fine to fix it later. This is not a critical issue.
